### ISSUE TITLE
CertificateOptions getattr calls have wrong attribute names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8315]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#8315|
|Type|defect|
|Created|2016-05-03 00:13:54Z|
|Branch|https://github.com/twisted/twisted/tree/attribute-spaces-8315|

There's an extraneous space in some of the attribute names:
```
    _OP_CIPHER_SERVER_PREFERENCE = getattr(SSL, 'OP_CIPHER_SERVER_PREFERENCE ',
                                           0x00400000)
    _OP_SINGLE_ECDH_USE = getattr(SSL, 'OP_SINGLE_ECDH_USE ', 0x00080000)
```
Fortunately I think this is relatively harmless since the constants in question have never changed value.

<details><summary>Searchable metadata</summary>

```
trac-id__8315 8315
type__defect defect
reporter__mithrandi mithrandi
priority__normal normal
milestone__None None
branch__branches_attribute_spaces_8315 branches/attribute-spaces-8315
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1462234434729342 1462234434729342
changetime__1462590269246684 1462590269246684
version__None None
owner__mithrandi mithrandi

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_1"></a>|@mithrandi commented|
|-|-|

(In [47334]) Branching to attribute-spaces-8315.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @mithrandi|
|-|-|

Thanks for the fix.

-------

This was introduced in [#6801](https://github.com/twisted/twisted/issues/6801) https://github.com/twisted/twisted/commit/b8ed5f3fe7e05bbb6f000f9684ead06649bfabbe

------

This bug is hit when you use a pyopenssl version which does not export those values.

This affects Twisted when used with PyOpenSSL 0.13

See this report https://bugs.launchpad.net/pyopenssl/+bug/1244201T

And the code http://bazaar.launchpad.net/~exarkun/pyopenssl/master/view/head:/OpenSSL/SSL.py

```
>>> OpenSSL.__version__
>>> '0.13'
OpenSSL.SSL.OP_SINGLE_ECDH_USE
>>> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'OP_SINGLE_ECDH_USE'
```

Do you know in which version of pyOpenSSL were those constants introduced?

--------

Since this is a defect, it should have a .bugfix news fragment with an information about what was fixed.

---------

If we want to support pyopenssl 0.13, then I think that we need a  builder running this version of openssl.

In this way we can write a real test for this code.

-----

If we don't have a slave, but we want to support pyopenssl 0.13, then we need a test using a mocked OpenSSL.

------

I am not sure what is the policy for supporting the pyopenssl version.

If we no longer support pyopenssl 0.13, we can just remove this code.

--------

Please get in touch with Amber or Glyph to define the supported pyopenssl version and to see if we can get a slave with pyopenssl 0.13 ... and write the appropriate tests (real or fake)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi removed owner|
|-|-|

I'm not sure in which version of pyOpenSSL the constants were introduced.

I'm counting this as .misc instead of .bugfix since the correct behaviour occurs in all cases even without this patch; it just means the fallback value is *always* used, instead of using the constant value from PyOpenSSL. If the constant value were to change, then the behaviour would be incorrect, but in practice it's unlikely for this to happen since it would break ABI compatibility for anything linked against OpenSSL.

For the same reason, I don't know how to test this; the result with or without the patch is the same, the constant has the correct value in all cases.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @mithrandi|
|-|-|

True. Sorry for the noise.

Please merge. Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_6"></a>|@mithrandi set status to `closed`|
|-|-|

(In [47355]) Merge attribute-spaces-8315: Fix attribute names for SSL constants.

Author: mithrandi
Reviewer: adiroiban
Fixes: [#8315](https://github.com/twisted/twisted/issues/8315)

Some of the constant names had spaces, so the fallback was always used.
